Async Deletion of Previous Metadata and Statistics Files

danielhumanmod opened the pull request:
# Description
Currently, when handling DROP TABLE PURGE, Polaris only deletes data files and the current metadata in `TableCleanupTaskHandler`. However, other related files, such as historical metadata.json files, stat files, and others, are not deleted.

Iceberg has addressed this issue (see [Iceberg PR #9305](https://github.com/apache/iceberg/pull/9305)), but Polaris doesn’t adopt this solution due to factors like asynchronous deletion. This PR introduces additional deletions for the following files:

- Previous metadata files
- Statistic files

Fixes #289 

**Included Changes**
- Updated `TableCleanupTaskHandler.java` to schedule previous metadata and statistics deletion in batch
- Extended `ManifestFileCleanupTaskHandler` to delete metadata and statistics file batch
- Extended `TaskTestUtils.java` to write metadata with previous metadata and statistics files
- Add unit tests in `TableCleanupTaskHandlerTest.java` and `TableContentCleanupTaskHandlerTest`

**Recommended Review Order**:
`TableCleanupTaskHandler.java` ->  `ManifestFileCleanupTaskHandler.java` -> `TaskTestUtils.java` -> `TableCleanupTaskHandlerTest.java` and `ManifestFileCleanupTaskHandlerTest`

## Future Improvement
1. [Refactoring] Separate manifest files and metadata file batch, but with a common base class
2. [Feature] Add async deletion partition statistics files

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Unit Test:
- [x] TableCleanupTaskHandlerTest
- [x] ManifestFileCleanupTaskHandlerTest 

# Checklist:

Please delete options that are not relevant.

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -158,6 +160,22 @@ public boolean handleTask(TaskEntity cleanupTask) {
+            .collect(Collectors.toMap(ManifestFile::path, Function.identity(), (mf1, mf2) -> mf1))
+            .keySet()
+            .forEach(fileIO::deleteFile);
+        tableMetadata.snapshots().stream()
+            .map(Snapshot::manifestListLocation)
+            .forEach(fileIO::deleteFile);
```
I think we've already handled the manifest list files.

danielhumanmod replied:
> I think we've already handled the manifest files and manifest list files.

Thank you for pointing that out! I just notice that these are already handled in the `ManifestFileCleanupTaskHandler`. I’ve removed the redundant deletion logic in the latest commit.

collado-mike commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -158,6 +159,13 @@ public boolean handleTask(TaskEntity cleanupTask) {
           taskExecutor.addTaskHandlerContext(createdTask.getId(), CallContext.getCurrentContext());
         }
+
+        tableMetadata.previousFiles().stream()
+            .map(TableMetadata.MetadataLogEntry::file)
+            .forEach(fileIO::deleteFile);
+        tableMetadata.statisticsFiles().stream()
+            .map(StatisticsFile::path)
+            .forEach(fileIO::deleteFile);
```
We should either add retries for these or submit them as separate tasks. As is, if one of these files fails to delete, we'll retry the whole task and resubmit a new task for each manifest. If the manifests are already deleted when we retry, we'll get stuck in a retry loop.

danielhumanmod replied:
Thanks for the suggestion, Michael! I could refactor the process by deleting these files in a new task and executing them separately, similar to how we handle manifest file deletions. Does that align with what you recommended approach?

collado-mike replied:
Sorry, do you mean adding retry within the `TableCleanupTaskHandler.java`? If so, that's fine with me. I'd at least use an Executor to attempt the execution in parallel (i.e., so file2 isn't blocked during the delay between retries for file1).

TBH, I'm not very familiar with how Iceberg generates the statistic files. Does there tend to be one per snapshot? one per data file? If the latter, we could be talking about a very large number of files. If that's the case, I think submitting separate tasks with batches of files makes sense.

flyrain replied:
They are per snapshots, https://iceberg.apache.org/spec/#table-statistics.

collado-mike replied:
Is there only one statistics file per snapshot? The spec is not clear:
> A table can contain many statistics files associated with different table snapshots.

Unlike the partition statistics file, which is very clear:
> Each table snapshot may be associated with at most one partition statistics file.

flyrain replied:
Could be multiple statistics files as the following code shows:

https://github.com/apache/iceberg/blob/5439cbdb278232779fdd9a392bbf57f007f9bda0/core/src/main/java/org/apache/iceberg/TableMetadata.java#L535-L535

danielhumanmod replied:
> They are per snapshots, https://iceberg.apache.org/spec/#table-statistics.

> Could be multiple statistics files as the following code shows:
> 
> https://github.com/apache/iceberg/blob/5439cbdb278232779fdd9a392bbf57f007f9bda0/core/src/main/java/org/apache/iceberg/TableMetadata.java#L535-L535

Thank you for providing the context, @flyrain :)

> Sorry, do you mean adding retry within the `TableCleanupTaskHandler.java`? If so, that's fine with me. I'd at least use an Executor to attempt the execution in parallel (i.e., so file2 isn't blocked during the delay between retries for file1).
> 
> TBH, I'm not very familiar with how Iceberg generates the statistic files. Does there tend to be one per snapshot? one per data file? If the latter, we could be talking about a very large number of files. If that's the case, I think submitting separate tasks with batches of files makes sense.

@collado-mike Yes, if we go for retries, the logic will be within `TableCleanupTaskHandler.java`. But considering the information provided by Yufei, maybe separate tasks will be a more appropriate approach for stats file deletion?

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -97,22 +105,30 @@ public void testTableCleanup() throws IOException {
-          .hasSize(1)
-          .satisfiesExactly(
+          .hasSize(2)
+          .satisfiesExactlyInAnyOrder(
```
1. For the single metadata with single snapshot case, verify the ability to delete statistics file

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -318,21 +345,35 @@ public void testTableCleanupMultipleSnapshots() throws IOException {
-                  .getOrCreateMetaStoreManager(realmContext)
-                  .loadTasks(polarisCallContext, "test", 5)
-                  .getEntities())
+      List<PolarisBaseEntity> entities = metaStoreManagerFactory
```
2. For the single metadata with multiple snapshot case, verify the ability to delete all of the statistics files

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -354,14 +395,207 @@ public void testTableCleanupMultipleSnapshots() throws IOException {
+  }
+
+  @Test
+  public void testTableCleanupMultipleMetadata() throws IOException {
```
3. For the multiple metadata with multiple snapshot case, verify the ability to delete all of the statistics files and previous metadata

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TaskTestUtils.java:
```diff
@@ -62,10 +69,32 @@ static ManifestFile manifestFile(
+    return writeTableMetadata(fileIO, metadataFile, null, null, statisticsFiles, snapshots);
+  }
+
+  static TableMetadata writeTableMetadata(
```
Extend this API to write metadata with previous metadata and statistics file for each snapshot

danielhumanmod commented:
WIP: process files in batch and add unit test for `TableContentCleanupTask`

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -49,6 +51,7 @@ public class TableCleanupTaskHandler implements TaskHandler {
   private final TaskExecutor taskExecutor;
   private final MetaStoreManagerFactory metaStoreManagerFactory;
   private final Function<TaskEntity, FileIO> fileIOSupplier;
+  private static final int BATCH_SIZE = 10;
```
Tentatively set the BATCH_SIZE as 10, please feel free to let me know if there is a better option

danielhumanmod commented:
Hi @collado-mike @flyrain , I’ve refactored the deletion logic to perform asynchronous deletion in batches, with each batch handled in separate tasks. I’d appreciate any feedback or suggestions you might have :)

danielhumanmod commented:
> Hi @collado-mike @flyrain , I’ve refactored the deletion logic to perform asynchronous deletion in batches, with each batch handled in separate tasks. I’d appreciate any feedback or suggestions you might have :)

Hi @collado-mike @flyrain @eric-maynard I wanted to kindly request any feedback or suggestions whenever you have a chance. Thank you in advance for your time :)

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -158,11 +161,98 @@ public boolean handleTask(TaskEntity cleanupTask) {
+            fileIO,
+            cleanupTask,
+            metaStoreManager,
+            polarisCallContext);
```
We could merge these two method calls. How about collecting the file first, then dispatching them into tasks for deletion? We don't have to distinguish file types while collecting them.

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableContentCleanupTaskHandler.java:
```diff
@@ -0,0 +1,189 @@
+/**
+ * {@link TaskHandler} responsible for deleting previous metadata and statistics files of a table.
+ */
+public class TableContentCleanupTaskHandler implements TaskHandler {
```
This new class is largely same as class `ManifestFileCleanupTaskHandler`.  Can we try to merge it with `ManifestFileCleanupTaskHandler`?

flyrain reviewed (commented):
Thanks @danielhumanmod for working on it! left some coments.

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
-        LOGGER.error("Unable to delete data files from manifest {}", manifestFile.path(), e);
-        return false;
-      }
+  private boolean manifestFileCleanUp(
```
I refactored the deletion logic for the manifest and its files into a new method; no other changes were made

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
-        LOGGER.error("Unable to delete data files from manifest {}", manifestFile.path(), e);
-        return false;
-      }
+  private boolean cleanUpManifestFile(
```
For manifest and all its data, I just refactored the deletion logic into a new method; no other changes were made.

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
-    ManifestFile manifestFile = decodeManifestData(cleanupTask.getManifestFileData());
     TableIdentifier tableId = cleanupTask.getTableId();
     try (FileIO authorizedFileIO = fileIOSupplier.apply(task)) {
-
```
Combine manifest and content file cleanup in the same task, but ensure that each task processes only one type of data (either manifest or content files) at a time.

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -42,9 +42,11 @@
+ * it was already deleted by another task. 2. Table content files: It contains previous metadata and
+ * statistics files, which are grouped and deleted in batch
  */
 public class ManifestFileCleanupTaskHandler implements TaskHandler {
```
Renaming this task will triger lots of relevent changes. If a rename is needed, we may want to handle it in a separate PR to avoid too much changes

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandlerTest.java:
```diff
@@ -226,4 +233,228 @@ public void deleteFile(String location) {
       assertThatPredicate((String f) -> TaskUtils.exists(f, fileIO)).rejects(dataFile2Path);
     }
   }
+
```
The test cases in this file primarily verify theses cleanup:
1. straightforward cleanup
2. cleanup when files do not exist
3. cleanup with retries

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
The test cases in this file primarily verify theses scenarios:

1. Cleanup of a single metadata file with a single snapshot
2. Cleanup of a single metadata file across multiple snapshots
3. Cleanup of multiple metadata files across multiple snapshots (New)

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -76,7 +79,13 @@ public void testTableCleanup() throws IOException {
       TestSnapshot snapshot =
           TaskTestUtils.newSnapshot(fileIO, "manifestList.avro", 1, snapshotId, 99L, manifestFile);
       String metadataFile = "v1-49494949.metadata.json";
-      TaskTestUtils.writeTableMetadata(fileIO, metadataFile, snapshot);
```
1. Cleanup of a single metadata file with a single snapshot

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -303,7 +323,20 @@ public void testTableCleanupMultipleSnapshots() throws IOException {
               manifestFile1,
               manifestFile3); // exclude manifest2 from the new snapshot
       String metadataFile = "v1-295495059.metadata.json";
-      TaskTestUtils.writeTableMetadata(fileIO, metadataFile, snapshot, snapshot2);
```
2. Cleanup of a single metadata file across multiple snapshots

danielhumanmod commented on polaris-service/src/test/java/org/apache/polaris/service/task/TableCleanupTaskHandlerTest.java:
```diff
@@ -318,21 +351,181 @@ public void testTableCleanupMultipleSnapshots() throws IOException {
+    }
+  }
+
+  @Test
```
3. Cleanup of multiple metadata files across multiple snapshots (New)

danielhumanmod commented:
> Thanks @danielhumanmod for working on it! left some coments.

Thanks for the suggestions @flyrain , I've updated the code based on your comments and added some comment to help the code easier to understand. I’d appreciate any further feedback, and thanks for your time and patience!

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -154,15 +190,32 @@ public boolean handleTask(TaskEntity cleanupTask) {
     return false;
   }
+
+  private List<List<String>> getContentFileBatch(TableMetadata tableMetadata, FileIO fileIO) {
```
The name `contentFile` in Iceberg refers to a data file or a delete file. Let's pick another name to avoid confusion, as metadata files and statistic files are not content files. How about `metadataFile`?

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -154,15 +190,32 @@ public boolean handleTask(TaskEntity cleanupTask) {
+
+  private List<List<String>> getContentFileBatch(TableMetadata tableMetadata, FileIO fileIO) {
+    List<List<String>> result = new ArrayList<>();
+    List<String> contentFiles =
```
rename here

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -154,15 +190,32 @@ public boolean handleTask(TaskEntity cleanupTask) {
+        Stream.concat(
+                tableMetadata.previousFiles().stream().map(TableMetadata.MetadataLogEntry::file),
+                tableMetadata.statisticsFiles().stream().map(StatisticsFile::path))
+            .filter(file -> TaskUtils.exists(file, fileIO))
```
Do we need to check the file existence here? It will be checked at delete time, right?

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -142,8 +146,40 @@ public boolean handleTask(TaskEntity cleanupTask) {
-              .toList();
+                  });
+
+      Stream<TaskEntity> contentFileCleanupTasks =
```
-> `metadataFileCleanupTask`

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -102,7 +106,7 @@ public boolean handleTask(TaskEntity cleanupTask) {
       // read the manifest list for each snapshot. dedupe the manifest files and schedule a
       // cleanupTask
       // for each manifest file and its data files to be deleted
-      List<TaskEntity> taskEntities =
```
This method is quite long. Can we split it to multiple methods like this?
```
 var manifestFileCleanupTask = getTaskStream(...);
 var metadataFileCleanupTasks = getTaskEntityStream(cleanupTask, tableMetadata, fileIO, tableEntity, metaStoreManager, polarisCallContext);

var taskEntities = Stream.concat(manifestCleanupTasks, contentFileCleanupTasks).toList();
```

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -185,12 +236,18 @@ private CompletableFuture<Void> tryDelete(
   public static final class ManifestCleanupTask {
     private TableIdentifier tableId;
     private String manifestFileData;
+    private List<String> contentFileBatch;
```
rename here

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
+    List<String> validFiles =
+        contentFileBatch.stream().filter(file -> TaskUtils.exists(file, fileIO)).toList();
+    if (validFiles.isEmpty()) {
+      LOGGER
+          .atWarn()
+          .addKeyValue("contentFileBatch", contentFileBatch.toString())
+          .addKeyValue("tableId", tableId)
+          .log("Table content cleanup task scheduled, but the none of the file in batch exists");
+      return true;
```
Do we need to warn here if 1 out of 10 files doesn't exist?

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
+          .log("Exception detected during content file batch deletion", e);
+    }
+
+    LOGGER
+        .atInfo()
+        .addKeyValue("contentFileBatch", contentFileBatch.toString())
+        .addKeyValue("tableId", tableId)
+        .log("Content file batch deletion has completed");
```
We can either remove it or make it debug level. An info level log may not be necessary.

flyrain commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -68,58 +70,107 @@ public boolean canHandleTask(TaskEntity task) {
+          CompletableFuture.allOf(deleteFutures.toArray(new CompletableFuture[0]));
+      allDeletes.join();
+    } catch (Exception e) {
+      LOGGER
+          .atWarn()
+          .addKeyValue("contentFileBatch", contentFileBatch.toString())
+          .addKeyValue("tableId", tableId)
+          .log("Exception detected during content file batch deletion", e);
```
Should we throw here?

flyrain reviewed (commented):
Sorry for the delay, @danielhumanmod. Left some comments.

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/ManifestFileCleanupTaskHandler.java:
```diff
@@ -42,9 +42,11 @@
+ * files in a manifest and the manifest itself. Since data files may be present in multiple
+ * manifests across different snapshots, we assume a data file that doesn't exist is missing because
+ * it was already deleted by another task. 2. Table content files: It contains previous metadata and
+ * statistics files, which are grouped and deleted in batch
```
> Need to handle the same for partition statistics files also (in this PR or as a follow up) more details: [apache/iceberg#9409](https://github.com/apache/iceberg/pull/9409)

Thanks for your suggestion! Would it be okay to address this in a separate PR avoid making this one too large

ajantha-bhat replied:
works for me.

danielhumanmod commented on polaris-service/src/main/java/org/apache/polaris/service/task/TableCleanupTaskHandler.java:
```diff
@@ -154,15 +135,118 @@ public boolean handleTask(TaskEntity cleanupTask) {
     }
     return false;
   }
+
```
Extract previous manifest task creation into a new method, no new change added

collado-mike reviewed (changes requested):
Sorry for the slow iteration cycle, but before we check in, I'd like to split the task type. Splitting the task type now means we have freedom to change the handler class design later, but if we start storing different tasks with the same task type, we won't be able to split the handlers later on.

danielhumanmod commented:
> Sorry for the slow iteration cycle, but before we check in, I'd like to split the task type. Splitting the task type now means we have freedom to change the handler class design later, but if we start storing different tasks with the same task type, we won't be able to split the handlers later on.

Thanks @collado-mike for the suggestions, which are very helpful in improving the code quality. I’ve already split the task types in the cleanup task handler as the comments suggested and am very willing to refactor this in future PRs. Appreciate any additional feedback you may have

danielhumanmod commented:
Hi @flyrain, I have made some changes based on the suggestion of Michael, do you mind taking another look on this? Thanks!
